### ISSUE TITLE
API for managing route configuration and access levels

### ISSUE THREAD

**Issue by shogunpurple** [Budibase/budibase#841]
When I'm using the budibase builder, I need to store route configuration on the backend to map URLs to screens and access levels. This configuration will be consumed by the client side router and used to show the correct screens to the user based on their URL and role.

This work includes:
- Coming up with a data structure that can be consumed by the client side router
- Store this data structure in couchDB in the app database
- Refactor and improve the structure of access levels. The ability to provide custom access levels and have levels like "Logged in user" are important

**Comment by mike12345567** [comment]
### Thoughts on routing data structure

* Do we need routes to be their own data structures in the DB, we could have a single "route" document but then we enter the difficulty of matching up screen IDs to routes, when anything screen related is updated we'll need to make sure the route document is up to date as well. If these documents get out of sync with each other in anyway then whole app will break.
* The data structure that makes up the routes can be inferred from route information contained within each screen document.
* We can use a view to extract a `routing` property from each screen document - this would appear as:
```javascript
{
  "path": "/table/new",
  "accessLevel": "LOGGED_IN",
}
```
* A fetch API can be built from this view that would get all of the routing information from every screen, and construct it into a structure like:
```javascript
{
  "routes": {
    "/": {
      "screenId": "...",
      "fullpath": "/",
      "accessLevel": "LOGGED_IN",
      "subpaths": {
        "table": {
          "screenId": "...",
          "fullpath": "/table",
          "accessLevel": "LOGGED_IN",
          "subpaths": {
            "new": {
              "screenId": "...",
              "fullpath": "/table/new",
              "accessLevel": "POWER",
              "subpaths": {}
            },
            ":id": {
              screenId: "...",
              "fullpath": "/table/:id",
              "accessLevel: "LOGGED_IN",
              "subpaths": {},
            }
          }
        }
      }
    }
  }
}
```
* This on the face of it may appear like quite a complex structure, but it should be a repeating pattern the entire way down, meaning it should be easy to display in an expandable tree.
* My thinking is we use something like [Page.js](https://www.npmjs.com/package/page) for our router as I think it could be quite drop-in replacement compatible for the `screenRouter`, removing the need for it but functioning in a similar manner. If we were to do this then we can iterate through everything, finding all the `fullpath` properties (or maybe have a simple utility API for this) which then can be dropped into something like:
```
page("/", navigateTo(...))
page("/table", navigateTo(...))
page("/table/new", navigateTo(...))
page("/table/:id", navigateTo(...))
```
Where each `navigateTo` call takes us to the correct screen ID.

**Comment by shogunpurple** [comment]
I like this - using a view might be a good way to keep it completely separate and make for easy migrations of that view structure in the future rather than storing all the route config in a document that we would have to consistently mutate. This route configuration is inherently derived data coming from screens, after all. This will also not be an expensive view to build and manage in terms of size. It's unlikely you will have even close to a hundred screens in most apps.

A few questions:
- Does this account for different roles being able to see different versions of a screen?
- We might want `accessLevel` keys to point to IDs, since the names will be able to change. Would it be worth doing the same kind of `enrich` pattern we did with linked records to query everything required for routing?

For the router itself, I think we might be going down the svelte route, so something like svelte-spa-router or svelte-routing could be the one we use - although it follows pretty much the same kind of structure as `Page.js`.

**Comment by kevmodrome** [comment]
After our discussion yesterday we talked a bit about using svelte-routing but I think svelte-spa-router would actually be better here - we'd be getting some more Svelty features and it's very well maintained as well!

**Comment by kevmodrome** [comment]
I was thinking we could do something like this for access levels:

Screen.svelte
```svelte
<script>
import { getStructure } from 'some-api-lib?'
import standardComponents from 'standard-components' // This would be a map
const screenStructure = getStructure(screenId, accessLevel) // This gets the structure for the correct screen from the backend
</script>

{#each screenStructure as child}
<svelte:component this={standardComponents.get(child.component)} props={child.props} />
{/each}
```

Not sure how well this would blend with the back-end though?

One thing we'd need to do here though is to re-work the standard-components to all use `<svelte:component />` instead of the `bb.attachChildren()` but that's just some easy re-factoring (hopefully?)

**Comment by mike12345567** [comment]
@shogunpurple on your questions:
> Does this account for different roles being able to see different versions of a screen?
Not currently I actually forgot that, but its very easy to add, if instead of `accessLevel` and `screenId` being top level properties for each path level we have instead
```
"screens": {
  "DEFAULT": screenId
  "LOGGED_IN": screenId,
  "ADMIN": screenId
}
```
Where we map screenIds to particular access levels.
> We might want accessLevel keys to point to IDs, since the names will be able to change. Would it be worth doing the same kind of enrich pattern we did with linked records to query everything required for routing?
So my thinking was there would be another endpoint that returns access levels, thats something I haven't quite fleshed out yet, but there will be a separate API for creating, fetching and managing custom access levels. This will include a few default levels as well. In terms of custom access levels they will be managed as IDs but the built-ins might have snake case names like "LOGGED_IN" so they are easy to distinguish as things that are statically built into the system (as I'm sure there will be a few that users would always be creating otherwise).
We could easily go down the enrich route as well, including the access levels in the routes call just for ease rather than needing to fetch both of them at the same time.

**Comment by mike12345567** [comment]
@kevmodrome this is one thing I've been thinking about with the svelte routers, I'm not sure how any of the routers I've been looking at blend particularly well with the backend serving up all of the routing information for each app, anything that is based on static definitions could be difficult to manage.

I'd love if you could walk me through your thoughts on this as I've not really been able to work out how the backend can serve something like svelte-spa-router efficiently as I assume it entails essentially serving the routing information and then serving the screens on a per route basis as well.

**Comment by shogunpurple** [comment]
@kevmodrome moving to `svelte:component` or `<slot />` from `attachChildren` is likely to fall into the work that @aptkingston is doing.

**Comment by aptkingston** [comment]
So after a design session with @mike12345567 and @kevmodrome we've come up with a consensus of how to move forward.

TLDR: use `svelte-spa-router` to handle things like pushstate history, navigation by pushing or popping and telling us what route we're on. We'll use our own logic to decide which components to render based on our current route. 2 new API endpoints will be added to return the list of screens and list of routes the current user can handle, based on their cookie. We'll remove the concept of `clientFrontendDefinition.js`.

### Fetching routes and screens
The client app, using the new component SDK, will fetch a list of available routes and available screen definitions. The server will use their cookie to determine what they can access. These definitions will be stored by the component SDK and functions will be exposed to allow the client app to query them. The actual interface for this is TBD and will be in a separate issue,
The client app will fetch these definitions on at least 2 occasions - initially when the app is loaded, and after logging in.

### Config for svelte-spa-router
`svelte-spa-router` will be configured to serve a single component for every possible route in our definition, our main client app component (called Screen.svelte for this example).
The config for `svelte-spa-router` would then look something like this:

App.svelte
```html
<script>
  import Router from "svelte-spa-router"
  import { getRoutes } from "@budibase/component-sdk"
  import Screen from "./Screen.svelte"
 
  // Add all available routes
  const routes = getRoutes()
  let routeConfig = {}
  routes.forEach(route => {
    routeConfig[route.path] = Screen
  })
</script>

<Router routes={routeConfig} />
```

### Rendering client app routes
The client app can get the current route from `svelte-spa-router`, then use this to render the corresponding page. As routes are unique, we can basically use this as a substitute screen ID here to query the screen definitions and get the correct one.

Screen.svelte
```html
<script>
  import { getScreenDefinition } from "@budibase/component-sdk"
  import { location } from "svelte-spa-router"

  // Get the screen definition for the current route
  const screenDefinition = getScreenDefinition(location)
</script>

<svelte:component this={screenDefinition.component} props={screenDefinition.props} />
```

The above code hopefully provides a good enough overview of how this could work. Permissions and access levels are fully decided by the server, as it will use the cookie (if one exists) to determine the access level of the user and return only the corresponding pages and routes that they can see. This should prevent any sort of definition for either a route or a screen being visible by underprivileged users.

**Comment by mike12345567** [comment]
As a note, after a discussion it is believed the best course of action is for routing to be one to one with screens, each screen will have exactly one option for routing and access level.

This will be presented in the screens API when saving as:
```json
screen.routing = {
  "route": "/home",
  "accessLevelId": "ADMIN"
}
```

This means that each screen essentially has a specific access level and route attached to it, it will not be possible for a single screen to be arrived it via multiple access levels and routes.

**Comment by mike12345567** [comment]
Minor update to the routing structure:
```javascript
{
  "/sales": {
    "subpaths": {
      "sales": {
        "screens": {
          "ADMIN": adminScreenId,
        }
      },
      "sales/new": {
        "screens": {
          "BASIC": basicScreenId,
        }
      }
    }
  },
  "/": {
    "subpaths": {
      "/": {
        "screens": {}
      }
    }
  }
}
```

The idea here is that instead of the full branching tree as discussed before there will only be two levels, the top which is useful for the builder and then all the subpaths are the full path.
